docs/features/tagging/high-level-design.md

```diff
@@ -0,0 +1,151 @@
+  tags to **1 tag per feature**.
+- There will be an **UNIQUE INDEX** on the combination of **`projectId` and `LOWER(tag)`** in order to enforce
+  **capitalization uniqueness** of the tag names
+
+On the DB side, the size the `tag` is unbounded, but is bounded at the service level via validation by a configurable
+amount of
+max characters.
+
+Once the last feature with a given tag is removed or untagged, that tag won't be available to be selected as a tag
+anymore. Tags will be considered removed when there are no rows for a given `projectId` and `featureId` combination
+with a given tag name.
+
+**Capitalization uniqueness should be enforced at the service level**. When tagging/renaming a tag on a feature, a case
+insensitive check must be made first to see if there's an already existing tag, regardless of capitalization, and use
+that tag instead of the one provided by the user. For example, if an user tags a Feature with `Mogwai`, and other
+features for the same project are already tagged with `mogwai` instead, `mogwai` should be used as the new tag. However
+this capitalization uniqueness **should not be applied tag renames for a given project** as the user might want to
+change capitalization of an existing tag.
+
+Some other solutions considered were JSONB Arrays or an Array of text values, as an extra column on the `features`
+table, containing all the tags in plain text form. Although these are better performant solutions in some use cases (tag
+cloud generation, tag combination querying...), given the scope, budget and data volume, a more classic solution based
+on an external table with JOINs is enough for this case.
+
+More info on different tagging implementation approaches [here](http://www.databasesoup.com/2015/01/tag-all-things.html)
+
+These are some of the main criteria used for deciding on the external table solution:
+
+- **All tag managing operations are done per project**: this limits quite a bit the number rows that queries must
+  handle, since doing system wide queries are not contemplated.
+- **System wide features are non-taggable**.
+- **Low volume of feature entries**: each project will have features in the hundreds, low thousands at most.
+- **Low number of tag entries**: the initial requirement is to have a max of 1 tag per feature; it is plausible to
+  assume that even in the case that this limitation is expanded upon, the possible number of tags per feature would be
+  in the low tens.
+- **One tag per feature**: queries are further simplified because there won't be querying for tag combinations
+- **The original `features` table remains the same**
+- **Budget**
+
+## Workflow
+
+There are two distinct parts to the tagging system.
+
+### Feature tagging operations
+
+These endpoints deal with operating tags for an individual feature.
+
+- `PATCH /api/v1/geo-features/:featureId/tags`
+
+```typescript
+Payload
+{
+	projectId: string
```

Comment:
we can decide at implementation time, but I think we can safely leave out `projectId` here because we would need to look it up in any case, as "parent" project of the feature the user wants to edit, as we need the "real" `projectId` anyway in order to process ACLs (we can't rely on a user-supplied `projectId` for this).

Reply:
You mean doing a quick load of the feature metadata to retrieve its projectId right? Yeah, that could work too, and shouldn't be much of an overhead. We'll see when we get to it.